src/ripple/overlay/impl/Cluster.cpp

```diff
@@ -0,0 +1,136 @@
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#include <BeastConfig.h>
+#include <ripple/app/main/Application.h>
+#include <ripple/basics/Log.h>
+#include <ripple/core/Config.h>
+#include <ripple/core/TimeKeeper.h>
+#include <ripple/overlay/Cluster.h>
+#include <ripple/overlay/ClusterNode.h>
+#include <ripple/protocol/JsonFields.h>
+#include <ripple/protocol/RippleAddress.h>
+#include <ripple/protocol/tokens.h>
+#include <boost/regex.hpp>
+#include <memory.h>
+
+namespace ripple {
+
+Cluster::Cluster (beast::Journal j)
+    : j_ (j)
+{
+}
+
+boost::optional<std::string>
+Cluster::member (RippleAddress const& identity)
+{
+    std::lock_guard<std::mutex> _(lock_);
+
+    auto iter = nodes_.find (identity);
+    if (iter == nodes_.end ())
+        return boost::none;
+    return iter->name ();
+}
+
+bool
+Cluster::update (
+    RippleAddress const& identity,
+    std::string name,
+    std::uint32_t loadFee,
+    std::uint32_t reportTime)
+{
+    std::lock_guard<std::mutex> _(lock_);
+
+    // NIKB FIXME: We can't use auto here yet due to an
+    //             issue with libstdc++.
+    std::set<ClusterNode, Comparator>::iterator iter =
+        nodes_.find (identity);
+
+    if (iter != nodes_.end ())
+    {
+        if (reportTime <= iter->getReportTime())
+            return false;
+
+        if (name.empty())
+            name = iter->name();
+
+        iter = nodes_.erase (iter);
+    }
+
+    nodes_.emplace_hint (iter, identity, name, loadFee, reportTime);
+    return true;
+}
+
+void
+Cluster::for_each (
+    std::function<void(ClusterNode const&)> func)
+{
+    std::lock_guard<std::mutex> _(lock_);
+    for (auto const& ni : nodes_)
+        func (ni);
+}
+
+std::unique_ptr<Cluster>
+make_Cluster (Config const& config, beast::Journal j)
+{
+    static boost::regex const re (
+        "^"                         // start of line
+        "(?:\\s*)"                  // whitespace (optional)
+        "([a-zA-Z0-9]*)"            // Node identity
+        "(?:\\s*)"                  // whitespace (optional)
+        "(.*\\S*)"                  // <value>
+        "(?:\\s*)"                  // whitespace (optional)
+    );
+
+    auto cluster = std::make_unique<Cluster> (j);
+
+    for (auto const& n : config.CLUSTER_NODES)
+    {
+        boost::smatch match;
```

Comment:
This page (https://isocpp.org/blog/2014/04/gcc-4.9.0) suggests that regex in gcc is now up-to-snuff for us.